File: Everyday/p658.java
import java.util.ArrayList;
import java.util.List;

public class p658 {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        int n = arr.length;
        List<Integer> ans = new ArrayList<>();
        if (n < k)
            return ans;
        int right = binSearch(arr, x);
        int left = right - 1;
        while (k-- > 0) {
            if (left < 0)
                ++right;
            else if (right >= n || x - arr[left] <= arr[right] - x)
                --left;
            else
                ++right;
        }
        ++left;
        while (left < right) {
            ans.add(arr[left++]);
        }
        return ans;
    }

    int binSearch(int[] arr, int x) {
        int left = 0, right = arr.length-1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (arr[mid] >= x) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
}

File: Everyday/p658.py
from bisect import bisect_left
from turtle import left, right
from typing import List


class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        n = len(arr)
        if n < k:
            return []
        if n == k:
            return arr
        right = bisect_left(arr, x)
        left = right-1
        for _ in range(k):
            if left < 0:
                right += 1
            elif right >= n or x-arr[left] <= arr[right]-x:
                left -= 1
            else:
                right += 1
        return arr[left+1:right]


arr = [-5, -3, 3, 4, 5]
print(Solution().findClosestElements(arr, 4, 3))
